crates/red_knot/src/salsa_db/semantic/types.rs

```diff
@@ -0,0 +1,662 @@
+
+    fn infer_definitions(
+        &self,
+        mut definitions: impl Iterator<Item = ReachableDefinition>,
+        symbol_scope: GlobalId<ScopeId>,
+    ) -> DefinitionType {
+        let Some(first) = definitions.next() else {
+            return DefinitionType::Type(Type::Unknown);
+        };
+
+        let typing_scope = TypingScope::for_symbol_scope(self.db(), symbol_scope);
+        let types = self.types(typing_scope);
+
+        if let Some(second) = definitions.next() {
+            let elements: FxIndexSet<_> = [first, second]
+                .into_iter()
+                .chain(definitions)
+                .map(|definition| self.infer_definition(definition, types))
+                .collect();
+
+            // The fact that the interner is local to a body means that we can't reuse the same union type
+            // across different call sites. But that's something we aren't doing yet anyway. Our interner doesn't
+            // deduplicate union types that are identical.
+            DefinitionType::Union {
+                ty: UnionType { elements },
+                scope: typing_scope,
+            }
+        } else {
+            DefinitionType::Type(self.infer_definition(first, types))
+        }
+    }
+
+    /// Infers the type of a location definition.
+    fn infer_definition(
+        &self,
+        definition: ReachableDefinition,
+        definition_types: &TypeInference,
+    ) -> Type {
+        let ReachableDefinition::Definition(definition) = definition else {
+            return Type::Unbound;
+        };
+
+        definition_types
+            .local_definitions
+            .get(&definition)
+            .copied()
+            .unwrap_or(Type::Unbound)
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Default)]
+pub struct TypeInference {
+    module: Option<ModuleType>,
+    functions: IndexVec<FunctionTypeId, FunctionType>,
+    classes: IndexVec<ClassTypeId, ClassType>,
+    unions: IndexVec<UnionTypeId, UnionType>,
+    intersections: IndexVec<IntersectionTypeId, IntersectionType>,
+
+    local_definitions: FxHashMap<Definition, Type>,
+    public_symbol_types: FxHashMap<SymbolId, Type>,
+    expression_types: FxHashMap<ExpressionId, Type>,
+}
+
+impl TypeInference {
+    fn shrink_to_fit(&mut self) {
+        self.functions.shrink_to_fit();
+        self.classes.shrink_to_fit();
+        self.unions.shrink_to_fit();
+        self.intersections.shrink_to_fit();
+
+        self.public_symbol_types.shrink_to_fit();
+        self.expression_types.shrink_to_fit();
+        self.local_definitions.shrink_to_fit();
+    }
+
+    pub fn symbol_ty(&self, id: SymbolId) -> Type {
+        self.public_symbol_types
+            .get(&id)
+            .copied()
+            .unwrap_or(Type::Unknown)
+    }
+
+    pub fn expression_ty(&self, id: ExpressionId) -> Type {
+        self.expression_types
+            .get(&id)
+            .copied()
+            .unwrap_or(Type::Unknown)
+    }
+}
+
+enum DefinitionType {
```

Comment:
The enum is a lifetime hack. `infer_definition` takes `&self` as argument, so it can't intern a new union type. 

The fact that it is a readonly reference is important in `finish` where we iterate over `self.symbol_table`. 

```rust
        for symbol in self.symbol_table.symbol_ids_for_scope(self.enclosing_scope) {
            let definition_type = self.typing_context().infer_definitions(
                symbol_table
                    .definitions(symbol)
                    .iter()
                    .map(|definition| ReachableDefinition::Definition(*definition)),
                GlobalId::new(self.file, self.enclosing_scope),
            );

            public_symbol_types.insert(symbol, definition_type.into_type(&mut self.result));
        }
```

Taking a &mut wouldn't compile because Rust couldn't prove that the `symbol_table` doesn't get mutated (a method taking `&mut self` can mutate any field). By explicitly passing `&mut self.result` in `into_type` Rust can prove that `self.symbol_table` is never borrowed mutably